### PR TITLE
feat(core/inlines): make variables context aware

### PR DESCRIPTION
**Pull request by marcoscaceres**
This is a work in progress

### CODE REVIEW

**Comment by marcoscaceres** [comment]
@sidvishnoi would like to chat about this. I have some ideas and could use your help.

File: src/core/inline-idl-parser.js [hunk 1]
```diff
@@ -9,75 +9,180 @@
...
-    result.method = splitted.pop();
-    result.args = result.method.match(methodRegex)[1].split(/,\s*/);
-  }
-  if (splitted.length > 1 && !result.method) {
-    result.attribute = splitted.pop();
+  const methodRegex = /(\w+)\((.*)\)$/;
+  const slotRegex = /$\[\[(\w+)\]\]$/;
+  const attributeRegex = /^(\w+)$/;
+  const enumRegex = /^(\w+)\["([\w ]+)"\]$/;
+  const tokens = str.split(".");
```
Comment (sidvishnoi) [inline src/core/inline-idl-parser.js]: Lets split at `/\b\.\b/` to add support for variadic methods.

``` js
"foo.bar.baz(foo, ...bar)".split(".")
// ["foo", "bar", "baz(foo, ", "", "", "bar)"]
"foo.bar.baz(foo, ...bar)".split(/\b\.\b/)
// ["foo", "bar", "baz(foo, ...bar)"]
```

Comment (marcoscaceres) [reply src/core/inline-idl-parser.js]: Unfortunately, that one doesn't work well for:
"foo.bar.[[test]].baz(foo, ...bar)".split(/\b\.\b/)

We might need to split on method, then split the rest? Or so some kind of fancy look ahead, the split behind.

Comment (sidvishnoi) [reply src/core/inline-idl-parser.js]:
```js
const splitRegex = /(?<=\]\]|\b)\./
```
😎 🤓

**Comment by sidvishnoi** [comment]
@marcoscaceres Removing `{{{ Dictionary["member"] }}}` syntax would work fine for referencing local IDL as we will be matching it up with local `pre.idl` (maybe a Map), but it would break the external references (in core/xref) as we would have no way to tell the xref API that we're looking for a `dict-member`. Or am I missing something?

**Comment by marcoscaceres** [comment]
> @marcoscaceres Removing {{{ Dictionary["member"] }}} syntax would work fine for referencing local IDL as we will be matching it up with local pre.idl (maybe a Map), but it would break the external references (in core/xref) as we would have no way to tell the xref API that we're looking for a dict-member. Or am I missing something?

All IDL containers are uniquely named, so the distinction between dictionary and interface is kinda an optimization. So, if we make the the micro-syntax for both the same, then it only becomes a user detail that one is a dictionary and another is an interface. 

That is:

```
 {{{ DictionaryWhatever }}}
 {{{ DictionaryWhatever.memberWhatever }}}
```

File: src/core/xref.js [hunk 2]
```diff
@@ -101,7 +101,17 @@ function createXrefMap(elems) {
       specs.push(...refs);
     }
 
-    const types = [isIDL ? elem.dataset.xrefType || "_IDL_" : "_CONCEPT_"];
+    const types = [];
+    if (isIDL) {
+      if (elem.dataset.xrefType) {
+        types.push(...elem.dataset.xrefType.split("|"));
```
Comment (sidvishnoi) [inline src/core/xref.js]: this will let us support treating attribute and dict-member the same

File: tests/spec/core/xref-spec.js [hunk 3]
```diff
@@ -547,7 +547,7 @@ describe("Core — xref", () => {
       expect(link2.href).toEqual(
         expectedLinks.get("PermissionStatus.[[query]]")
       );
-      expect(link2.textContent).toEqual("[[query]]");
+      expect(link2.textContent).toEqual("query");
```
Comment (sidvishnoi) [inline tests/spec/core/xref-spec.js]: need a thought here

File: src/core/inline-idl-parser.js [hunk 4]
```diff
@@ -9,77 +9,196 @@
...
+  const { parent, identifier } = details;
+  const { idlType } = parent;
+  const html = hyperHTML`.<a
+      class="respec-idl-xref"
+      data-xref-type="attribute|dict-member"
+      data-link-for="${idlType}">${identifier}</a>`;
+  return html;
+}
+
+// Method: .identifier(arg1, arg2, ...)
```
Comment (sidvishnoi) [inline src/core/inline-idl-parser.js]:
```suggestion
// Method: .identifier(arg1, arg2, ...), identifier(arg1, arg2, ...)
```

File: src/core/inline-idl-parser.js [hunk 5]
```diff
@@ -9,77 +9,196 @@
...
+    class="respec-idl-xref"
+    data-xref-type="${type}"
+    data-link-for="${idlType}"
+    data-lt="${searchText}"
+    >${identifier}</a>(${[argsText]})`;
+  return html;
+}
+
+// Enum: "enum value"
+function renderEnum(details) {
```
[the remark this answers is not in the document]
Comment (marcoscaceres) [reply src/core/inline-idl-parser.js]: hmm... we should support this.

[the remark this answers is not in the document]
Comment (marcoscaceres) [reply src/core/inline-idl-parser.js]: {{{ "some value" }}}

**Comment by marcoscaceres** [comment]
I think we should beta test this with Payment Request a bit.

File: src/core/inline-idl-parser.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (sidvishnoi) [reply src/core/inline-idl-parser.js]: `const enumRegex = /^(\w+)\["([\w ]+)"\]$/;` seems to be matching something else

File: src/core/inline-idl-parser.js [hunk 6]
```diff
@@ -9,77 +9,215 @@
...
 function parseInlineIDL(str) {
-  const result = Object.create(null);
-  const splitted = str.split(idlSplitRegex);
-  if (methodRegex.test(splitted[splitted.length - 1])) {
-    result.method = splitted.pop();
-    result.args = result.method.match(methodRegex)[1].split(/,\s*/);
-  }
-  if (splitted.length > 1 && !result.method) {
-    result.attribute = splitted.pop();
+  const methodRegex = /(\w+)\((.*)\)$/;
```
Comment (sidvishnoi) [inline src/core/inline-idl-parser.js]: we should move these regex outside this function maybe

**Pull request closed without merging**